File: njli/physics/shape/PhysicsShape.cpp
//
//  PhysicsShape.cpp
//  JLIGameEngineTest
//
//  Created by James Folk on 11/22/14.
//  Copyright (c) 2014 James Folk. All rights reserved.
//

#include "PhysicsShape.h"
#include "JLIFactoryTypes.h"
#include "World.h"
//#include "PhysicsShapeBuilder.h"
#include "PhysicsBody.h"
#include "btCollisionShape.h"
#include "btMotionState.h"
#include "btRigidBody.h"
#define FORMATSTRING "{\"njli::PhysicsShape\":[]}";

namespace njli
{
  PhysicsShape::PhysicsShape() : AbstractFactoryObject(this) {}

  PhysicsShape::PhysicsShape(const AbstractBuilder &builder)
      : AbstractFactoryObject(this)
  {
    // TODO: Implement...
  }

  PhysicsShape::PhysicsShape(const PhysicsShape &copy)
      : AbstractFactoryObject(this)
  {
    // TODO: Implement...
  }

  PhysicsShape::~PhysicsShape() {}

  PhysicsShape &PhysicsShape::operator=(const PhysicsShape &rhs)
  {
    if (this != &rhs)
      {
        setCollisionShape(*(rhs.getCollisionShape()));
        //            m_PropertyChanged = true;
      }
    return *this;
  }

  PhysicsShape *PhysicsShape::create(u32 type)
  {
    //        SDL_assert(type == JLI_OBJECT_TYPE_PhysicsShapeBox);

    return dynamic_cast<PhysicsShape *>(
        World::getInstance()->getWorldFactory()->create(type));
  }

  void PhysicsShape::destroy(PhysicsShape *object)
  {
    if (object)
      {
        World::getInstance()->getWorldFactory()->destroy(object);
      }
  }

  void PhysicsShape::load(PhysicsShape &object, lua_State *L, int index)
  {
    // Push another reference to the table on top of the stack (so we know
    // where it is, and this function can work for negative, positive and
    // pseudo indices
    lua_pushvalue(L, index);
    // stack now contains: -1 => table
    lua_pushnil(L);
    // stack now contains: -1 => nil; -2 => table
    while (lua_next(L, -2))
      {
        // stack now contains: -1 => value; -2 => key; -3 => table
        // copy the key so that lua_tostring does not modify the original
        lua_pushvalue(L, -2);
        // stack now contains: -1 => key; -2 => value; -3 => key; -4 => table
        const char *key = lua_tostring(L, -1);
        //            const char *value = lua_tostring(L, -2);
        if (lua_istable(L, -2))
          {
            PhysicsShape::load(object, L, -2);
          }
        else
          {
            if (lua_isnumber(L, index))
              {
                double number = lua_tonumber(L, index);
                printf("%s => %f\n", key, number);
              }
            else if (lua_isstring(L, index))
              {
                const char *v = lua_tostring(L, index);
                printf("%s => %s\n", key, v);
              }
            else if (lua_isboolean(L, index))
              {
                bool v = lua_toboolean(L, index);
                printf("%s => %d\n", key, v);
              }
            else if (lua_isuserdata(L, index))
              {
                //                    swig_lua_userdata *usr;
                //                    swig_type_info *type;
                //                    assert(lua_isuserdata(L,index));
                //                    usr=(swig_lua_userdata*)lua_touserdata(L,index);
                //                    /* get data */
                //                    type = usr->type;
                //                    njli::AbstractFactoryObject *object =
                //                    static_cast<njli::AbstractFactoryObject*>(usr->ptr);
                //                    printf("%s => %d:%s\n", key,
                //                    object->getType(),
                //                    object->getClassName());
              }
          }
        // pop value + copy of key, leaving original key
        lua_pop(L, 2);
        // stack now contains: -1 => key; -2 => table
      }
    // stack now contains: -1 => table (when lua_next returns 0 it pops the key
    // but does not push anything.)
    // Pop table
    lua_pop(L, 1);
    // Stack is now the same as it was on entry to this function
  }

  u32 PhysicsShape::type() { return JLI_OBJECT_TYPE_PhysicsShape; }

  btRigidBody *
  PhysicsShape::createRigidBodyWithShape(f32 mass, btMotionState *motionState,
                                         const btVector3 &localInertia)
  {
    btRigidBody *rb =
        new btRigidBody(mass, motionState, getCollisionShape(), localInertia);
    return rb;
  }

  btVector3 PhysicsShape::calculateLocalInertia(btScalar mass) const
  {
    btVector3 localInertia(0, 0, 0);
    if (mass > 0.0f)
      {
        getCollisionShape()->calculateLocalInertia(mass, localInertia);
      }
    return localInertia;
  }

  //    void PhysicsShape::enablePropertyChange(bool enable)
  //    {
  //        m_PropertyChanged = enable;
  //    }
  //    bool PhysicsShape::isPropertyChanged()const
  //    {
  //        return m_PropertyChanged;
  //    }

  void PhysicsShape::setLocalScaling(const btVector3 &scaling)
  {
    getCollisionShape()->setLocalScaling(scaling);
  }

  void PhysicsShape::getAabb(const btTransform &t, btVector3 &aabbMin,
                             btVector3 &aabbMax) const
  {
    getCollisionShape()->getAabb(t, aabbMin, aabbMax);
  }

  //    void PhysicsShape::applyPhysicsShape()
  //    {
  //        PhysicsBody *body = getParent();
  //        if(body)
  //        {
  //            Node *node = body->getParent();
  //            if(node)
  //            {
  //                node->applyPhysicsShape();
  //            }
  //        }
  //    }

  PhysicsBody *PhysicsShape::getParent()
  {
    // return dynamic_cast<PhysicsBody*>(getParent());
    return NULL;
  }
  const PhysicsBody *PhysicsShape::getParent() const
  {
    // return dynamic_cast<const PhysicsBody*>(getParent());
    return NULL;
  }
}

File: njli/platform/cmake.in/ldoc.in/PhysicsShape.lua

----
-- @file PhysicsShape


---- Brief description.
-- <#Description#>
-- @return <#return value description#>
function PhysicsShape:calculateSerializeBufferSize()
end

---- Brief description.
-- <#Description#>
-- @param btSerializer <#btSerializer description#>
-- @return <#return value description#>
function PhysicsShape:serialize(btSerializer)
end

---- Brief description.
-- <#Description#>
-- @return <#return value description#>
function PhysicsShape:getClassName()
end

---- Brief description.
-- <#Description#>
-- @return <#return value description#>
function PhysicsShape:getType()
end

---- Brief description.
-- <#Description#>
-- @param mass <#mass description#>
-- @param motionState <#motionState description#>
-- @param localInertia <#0 description#>
-- @return <#return value description#>
function PhysicsShape:createRigidBodyWithShape(mass, motionState, localInertia)
end

---- Brief description.
-- <#Description#>
-- @param mass <#mass description#>
-- @return <#return value description#>
function PhysicsShape:calculateLocalInertia(mass)
end

---- Brief description.
-- <#Description#>
-- @return <#return value description#>
function PhysicsShape:getNumVertices()
end

---- Brief description.
-- <#Description#>
-- @return <#return value description#>
function PhysicsShape:getNumEdges()
end

---- Brief description.
-- <#Description#>
-- @param i <#i description#>
-- @param pa <#pa description#>
-- @param pb <#pb description#>
-- @return <#return value description#>
function PhysicsShape:getEdge(i, pa, pb)
end

---- Brief description.
-- <#Description#>
-- @param i <#i description#>
-- @param vtx <#vtx description#>
-- @return <#return value description#>
function PhysicsShape:getVertex(i, vtx)
end

---- Brief description.
-- <#Description#>
-- @return <#return value description#>
function PhysicsShape:getNumPlanes()
end

---- Brief description.
-- <#Description#>
-- @param planeNormal <#planeNormal description#>
-- @param planeSupport <#planeSupport description#>
-- @param i <#i description#>
-- @return <#return value description#>
function PhysicsShape:getPlane(planeNormal, planeSupport, i)
end

---- Brief description.
-- <#Description#>
-- @param pt <#pt description#>
-- @param tolerance <#tolerance description#>
-- @return <#return value description#>
function PhysicsShape:isInside(pt, tolerance)
end

---- Brief description.
-- <#Description#>
-- @param scaling <#scaling description#>
-- @return <#return value description#>
function PhysicsShape:setLocalScaling(scaling)
end

---- Brief description.
-- <#Description#>
-- @param t <#t description#>
-- @param aabbMin <#aabbMin description#>
-- @param aabbMax <#aabbMax description#>
-- @return <#return value description#>
function PhysicsShape:getAabb(t, aabbMin, aabbMax)
end

---- Brief description.
-- <#Description#>
-- @param type <
-- @return <#return value description#>
function NJLI.PhysicsShape.create(type)
end

---- Brief description.
-- <#Description#>
-- @param object <#object description#>
-- @return <#return value description#>
function NJLI.PhysicsShape.destroy(object)
end

---- Brief description.
-- <#Description#>
-- @param object <#object description#>
-- @param L <#L description#>
-- @param stack_index <#stack_index description#>
-- @return <#return value description#>
function NJLI.PhysicsShape.load(object, L, stack_index)
end

---- Brief description.
-- <#Description#>
-- @return <#return value description#>
function NJLI.PhysicsShape.type()
end

